### ISSUE TITLE
Fix Lithops zombie processes issue

### ISSUE THREAD

**Issue by LachlanStuart** [metaspace2020/metaspace#795]
This task has 2 parts: First mitigate the issue on production so that Lithops doesn't require manual intervention every day, then test with the latest Lithops version and work with the Lithops team to ensure is problem is fixed at the Lithops level.

# Problem

Under certain conditions, Lithops' background processes aren't cleaned up, which sometimes prevents the main process from exiting, and sometimes allows these background processes to stay alive forever. 

If they prevent the main process from exiting, Supervisor doesn't restart the daemon, meaning the daemons stop processing annotation messages and aren't restarted.

If processes stay alive in the background, they [continue making COS API requests](https://github.com/lithops-cloud/lithops/issues/395) which can become quite expensive over time.

# Part 1: Mitigation

I currently have to manually restart Lithops and clean up the background processes every day on production, so putting a workaround in place should be a high priority. If I don't do this every day, annotation stops completely.

The fastest way to a solution is to hotfix METASPACE so that it kills all child processes before exiting in the `cb_stop_daemons` handler in `run_sm_daemons.py`.

# Part 2: Permanent fix

I've seen this problem occur under a variety of circumstances. Some may be fixed already:

1. A function is killed due to using too much memory
2. A function is killed for running too long
3. A code/config error causes the Lithops code that launches the user code to fail (this prevents it from writing a status file)
4. A function exits due to an exception while there are still functions queued to start (the invoker stops starting new functions, but the job monitor waits forever for these functions to run)
5. A function fails to start due to the Lithops runtime cache and IBM Cloud going out of sync. Lithops caches which runtimes have been configured, but it's possible for these to be changed without the cache being updated.
6. An function running in `ibm_vpc` fails to complete due to an error in the "proxy" or "runner" services running on the VM (can be simulated by SSHing into the instance and killing Python processes after execution has started)

I've made a script to test cases 1, 2, 3 & 4: https://gist.github.com/LachlanStuart/a8f4b75d3aecbdb07af5e5b9d479ada6 . For each test case, the call to `fexec.get_result(future)` should always throw an exception, but the following asserts should never fail. 5 probably can be skipped, but 6 will need to be manually tested.

Please test this with the latest Lithops version or [master branch](https://github.com/lithops-cloud/lithops). Try it with the 3 executor backends/modes that we care about: `ServerlessExecutor` with the `ibm_cf` backend, `StandaloneExecutor` with `ibm_vpc`, and `ServerlessExecutor` with `code_engine`. Work with the Lithops team to get all discovered bugs fixed.

**Comment by gilv** [comment]
@LachlanStuart I just used Lithops master branch v 2.3.1b1, where i wrote a function

```
def oom_test(val):
    
    if (val==3):
        size = 1024 * 1024 * 256
        data = "d" * (int(size / 2) - 1)
    return val
```

I then submitted 

```
iterdata = [1,2,3,4,5]
ce_runtime = "ibmfunctions/lithops-ce-v385-1:231b1"
pw = lithops.FunctionExecutor(mode = 'serverless', backend='code_engine', runtime=ce_runtime, runtime_memory=128, log_level='DEBUG')
futures = pw.map(oom_test, iterdata)
print (pw.get_result(futures))
pw.clean()
```

Got

```
2021-04-07 05:42:44,178 [INFO] lithops.invokers -- ExecutorID bd121c-0 | JobID M000 - View execution logs at /private/var/folders/nw/kg5kt9150nj48738r5wz9ly80000gn/T/lithops/logs/bd121c-0-M000.log
2021-04-07 05:42:44,178 [INFO] lithops.executors -- ExecutorID bd121c-0 - Getting results
2021-04-07 05:42:49,709 [DEBUG] lithops.future -- ExecutorID bd121c-0 | JobID M000 - Got status from call 00001 - Activation ID: 7c7db9e3d72f - Time: 0.52 seconds
2021-04-07 05:42:49,878 [DEBUG] lithops.future -- ExecutorID bd121c-0 | JobID M000 - Got output from call 00001 - Activation ID: 7c7db9e3d72f
2021-04-07 05:43:17,524 [DEBUG] lithops.future -- ExecutorID bd121c-0 | JobID M000 - Got status from call 00003 - Activation ID: f715dc6516b1 - Time: 0.21 seconds
2021-04-07 05:43:17,532 [DEBUG] lithops.future -- ExecutorID bd121c-0 | JobID M000 - Got status from call 00004 - Activation ID: 193433cb4e83 - Time: 0.34 seconds
2021-04-07 05:43:17,554 [DEBUG] lithops.future -- ExecutorID bd121c-0 | JobID M000 - Got status from call 00000 - Activation ID: 250715f16849 - Time: 0.96 seconds
2021-04-07 05:43:17,673 [DEBUG] lithops.future -- ExecutorID bd121c-0 | JobID M000 - Got output from call 00004 - Activation ID: 193433cb4e83
2021-04-07 05:43:17,685 [DEBUG] lithops.future -- ExecutorID bd121c-0 | JobID M000 - Got output from call 00003 - Activation ID: f715dc6516b1
2021-04-07 05:43:17,696 [DEBUG] lithops.future -- ExecutorID bd121c-0 | JobID M000 - Got output from call 00000 - Activation ID: 250715f16849
2021-04-07 05:43:17,697 [WARNING] lithops.future -- ExecutorID bd121c-0 | JobID M000 - There was an exception - Activation ID: 1661f3c25624
2021-04-07 05:43:17,697 [WARNING] lithops.future -- Exception: MemoryError - Function exceeded maximum memory and was killed
```

There was no stall issue. Is there other test you have so i can reproduce the stall issue on OOM ?

**Comment by LachlanStuart** [comment]
@gilv Please test with this script: https://gist.github.com/LachlanStuart/a8f4b75d3aecbdb07af5e5b9d479ada6 
(You'll need to change line 41 to load your lithops config instead of ours. In any case, make sure that `lithops.workers` is set to 1)

It tests 4 different scenarios, and more importantly, makes sure that all background tasks have been cleaned up

**Comment by gilv** [comment]
@LachlanStuart this is VPC related or CE related?

**Comment by LachlanStuart** [comment]
@gilv Please test both. I've seen these issues with both.

**Comment by gilv** [comment]
@LachlanStuart this script suppose to run in loop over all tests? it's not

```
Testing test_out_of_memory
2021-04-07 11:08:23,636 [INFO] lithops.config -- Lithops v2.3.1b1
2021-04-07 11:08:24,530 [INFO] lithops.storage.backends.ibm_cos.ibm_cos -- IBM COS Storage client created - Region: us-east
/Users/gilv/.pyenv/versions/3.8.5/lib/python3.8/site-packages/urllib3/connectionpool.py:981: InsecureRequestWarning: Unverified HTTPS request is being made to host 'iam.cloud.ibm.com'. Adding certificate verification is strongly advised. See: https://urllib3.readthedocs.io/en/latest/advanced-usage.html#ssl-warnings
  warnings.warn(
2021-04-07 11:08:28,005 [INFO] lithops.serverless.backends.code_engine.code_engine -- IBM Code Engine client created - Region: us-south
2021-04-07 11:08:28,005 [INFO] lithops.executors -- Serverless Executor created with ID: 52a4d1-0
2021-04-07 11:08:28,005 [INFO] lithops.invokers -- ExecutorID 52a4d1-0 | JobID M000 - Selected Runtime: ibmfunctions/lithops-ce-v385-1:231b1 - 256MB 
2021-04-07 11:08:28,010 [INFO] lithops.job.job -- ExecutorID 52a4d1-0 | JobID M000 - Uploading function and data - Total: 736.0B
2021-04-07 11:08:30,498 [INFO] lithops.invokers -- ExecutorID 52a4d1-0 | JobID M000 - Starting function invocation: test_out_of_memory() - Total: 3 activations
2021-04-07 11:08:31,371 [INFO] lithops.serverless.backends.code_engine.code_engine -- Job definition ibmfunctions--lithops-ce-v385-1--231b1--256mb not found (404)
2021-04-07 11:08:33,060 [INFO] lithops.invokers -- ExecutorID 52a4d1-0 | JobID M000 - View execution logs at /private/var/folders/nw/kg5kt9150nj48738r5wz9ly80000gn/T/lithops/logs/52a4d1-0-M000.log
2021-04-07 11:08:33,061 [INFO] lithops.executors -- ExecutorID 52a4d1-0 - Getting results

Lithops threw exception as expected: Function exceeded maximum memory and was killed
2021-04-07 11:08:42,050 [INFO] lithops.executors -- ExecutorID 52a4d1-0 - Cleaning temporary data
Traceback (most recent call last):
  File "/Users/gilv/Dev/ibm-pywren-internal/embl-tests/test_lithops_error_handling.py", line 72, in <module>
    assert not child_processes
AssertionError
```

**Comment by LachlanStuart** [comment]
@gilv I haven't had time to make this into a proper test suite, so if you want to run all the tests even though some fail, you may have to manually select tests by changing the list in line 46. This exception:
```
    assert not child_processes
AssertionError
```
Means that there is still some background Lithops process running (i.e. the test failed). 

In the past these background processes have always been due to a job monitor process that never stopped. [This bug was previously reported](https://github.com/lithops-cloud/lithops/issues/395) and fixed, but I believe it has regressed since them. These processes are a problem because they continue making COS requests forever and eventually lead to a large bill due to making many Class A API requests over a long period of time. One process isn't a big deal, but on our production server we process 20 datasets per day, so even if it only happens 10% of the time, that's still 14 new processes per week polling for status files that are never written, which quickly adds up to a lot of COS requests.

**Comment by gilv** [comment]
@LachlanStuart you think those processing still access COS, even Lithops reported error? I think in the past they accessed COS, because Lithops was stuck on get result, but now it's not stuck..so those processes do nothing

**Comment by gilv** [comment]
@LachlanStuart i will check it...what they do

**Comment by JosepSampe** [comment]
`ibm_vpc` or `code_engine` are job-like backends, so the unique subprocess belonging to Lithops is the "cleaner", responsible to delete the temp data in background at the end of an execution. It is spawned [here](https://github.com/lithops-cloud/lithops/blob/master/lithops/executors.py#L600) using the `subprocess.Popen` method. So in the [test script](https://gist.github.com/LachlanStuart/a8f4b75d3aecbdb07af5e5b9d479ada6) you should add a minimum sleep after `fexec.clean()` to give it time to finish (as it is not controlled by the multprocessing library).

In 'ibm_cf' the subprocesess are the `invoker` and the `job_monitor`, both are killed after an exception.

**Comment by gilv** [comment]
@JosepSampe I used

```
    fexec.clean()
    del fexec
    gc.collect()

    # Make sure all processes created through multiprocessing are cleaned up
    import time
    time.sleep(60*3)
    for child in multiprocessing.active_children():
 ```
and it still failed 
```
Traceback (most recent call last):
  File "/Users/gilv/Dev/ibm-pywren-internal/embl-tests/test_lithops_error_handling.py", line 74, in <module>
    assert not child_processes
AssertionError
```

**Comment by JosepSampe** [comment]
@gilv The subprocess.Popen() command keeps an "sh" terminated process until de main program exits, so you have to get rid of it with: `child_processes = [cp for cp in child_processes if cp.name() != 'sh']` before the last assert

**Comment by gilv** [comment]
@JosepSampe i run it from my Eclipse...

```
    # Make sure all processes created through subprocess are cleaned up
    child_processes = psutil.Process().children(recursive=True)
    child_processes = [cp for cp in child_processes if cp.name() != 'sh']
    for cp in child_processes:
        print (cp.name())
```

```
2021-04-07 22:23:16,266 [INFO] lithops.executors -- ExecutorID aa8dc6-0 - Getting results

Lithops threw exception as expected: Function exceeded maximum memory and was killed
2021-04-07 22:23:22,861 [INFO] lithops.executors -- ExecutorID aa8dc6-0 - Cleaning temporary data
python3.8
python3.8
Traceback (most recent call last):
  File "/Users/gilv/Dev/ibm-pywren-internal/embl-tests/test_lithops_error_handling.py", line 77, in <module>
    assert not child_processes
AssertionError
```

**Comment by JosepSampe** [comment]
@gilv Seems you still have 2 python processes. Can you print child_processes var?

**Comment by gilv** [comment]
@JosepSampe 

```
2021-04-07 22:34:31,979 [INFO] lithops.executors -- ExecutorID 33ed1f-0 - Getting results

Lithops threw exception as expected: Function exceeded maximum memory and was killed
2021-04-07 22:35:41,587 [INFO] lithops.executors -- ExecutorID 33ed1f-0 - Cleaning temporary data
[psutil.Process(pid=29969, name='python3.8', status='running', started='22:34:32'), psutil.Process(pid=29971, name='python3.8', status='zombie', started='22:35:42')]
python3.8
python3.8
Traceback (most recent call last):
  File "test_lithops_error_handling.py", line 78, in <module>
    assert not child_processes
AssertionError
```

**Comment by JosepSampe** [comment]
@gilv May they are started by eclipse. Could you try running the script directly from console?

Also, after `print(child_processes)`, add a `time.sleep(60)` and then check the `pid` you see in the print by running in console: `ps aux | grep <pid>`, this way you will see the full python command

**Comment by gilv** [comment]
@JosepSampe it also happens from the console....

```
 ps aux | grep 30112
gilv             30197   0.0  0.0  4259000     12 s014  S+   10:51PM   0:00.00 grep 30112
gilv             30112   0.0  0.1  4246468   8412 s013  S+   10:47PM   0:00.05 /Users/gilv/.pyenv/versions/3.8.5/bin/python -c from multiprocessing.resource_tracker import main;main(10)
```

**Comment by JosepSampe** [comment]
@gilv This seems to be something related to Mac. In Linux this process does not appear. What is the other process in `ps aux`?

**Comment by gilv** [comment]
@JosepSampe the other one was zombie... not even sure what it means. the ps aux returned nothing on 

```
psutil.Process(pid=30113, name='python3.8', status='zombie', started='22:48:00')
```

**Comment by JosepSampe** [comment]
And what if you comment this line https://github.com/lithops-cloud/lithops/blob/master/lithops/executors.py#L600 ?

**Comment by gilv** [comment]
@JosepSampe when this line commented, no zombie processes...only this one
```
gilv             30503   0.0  0.1  4274784   8732   ??  S    11:03PM   0:00.06 /Users/gilv/.pyenv/versions/3.8.5/bin/python3.8 -c from multiprocessing.resource_tracker import main;main(10)
```

and you think it'a mac issue, right?

**Comment by JosepSampe** [comment]
I wonder what might be this process

`python3.8 -c from multiprocessing.resource_tracker import main;main(10)`

It is not directly started by Lithops. Which backend are you using?

**Comment by gilv** [comment]
@JosepSampe code engine

**Comment by JosepSampe** [comment]
In `ibm_vpc ` and `code_engine` the unique process that Lithops starts is [this one](https://github.com/lithops-cloud/lithops/blob/master/lithops/executors.py#L600) you already commented.

This process you see might be related to python itself. Can you comment the lithops code (ServerlessExecutor, map, get_result, clean) in the script and check if it still appears?

**Comment by gilv** [comment]
@JosepSampe if i remark 
```
    '''
    fexec= FunctionExecutor(mode = 'serverless', backend='code_engine',runtime='ibmfunctions/lithops-ce-v385-1:231b1')
    future = fexec.map(test_out_of_memory, [(), (), ()], runtime_memory=256)
    try:
        fexec.get_result(future)
    except Exception as ex:
        print(f'Lithops threw exception as expected: {ex}')

    # Clean up the Lithops instance & ensure everything is garbage collected
    fexec.clean()
    del fexec
    '''
```
 then 
```
print (child_processes)
```
prints
```
[]
``` ;)

**Comment by JosepSampe** [comment]
@gilv Interesting... And what if you use ibm_cf instead of code_engine?

**Comment by gilv** [comment]
@JosepSampe the same processes when i am using cloud functions

**Comment by gilv** [comment]
@LachlanStuart @sergii-mamedov  i tested all the script (all tests ) with Lithops master branch and runtime `ibmfunctions/lithops-ce-v385-1:231b1`. I used code engine backend. Script run without issues, no stall issues at all.  So i think the stall issues on OOM or other things are not exists in the latest Lithops. 

Related 2 processes we see running (see my discussion with @JosepSampe ) - those processes has no meaning, unclear yet who start them..may be my mac, may be eclipse, may be pyenv, etc. for SURE this process 
```
python3.8 -c from multiprocessing.resource_tracker import main;main(10)
```
doesn't pull COS, which was your main concern.

**Comment by JosepSampe** [comment]
@LachlanStuart We recently released a new Lithops version (2.3.3) with completely new monitoring system. New Lithops version does not use multiprocessing Processes anymore. Now everything are daemonic threads, so this should't prevent your main program from exiting. Moreover, new monitoring system substantially reduces the number of COS requests. In addition, you will see now a periodic debug message stating the `pending`, `running` and `done` calls, so its is now easier to identify if the job monitor is working or not. Now, it also ensures that when an exception is produced, both the invoker and the job monitor are stopped, so the job monitor should not wait forever for not invoked calls.

**Comment by LachlanStuart** [comment]
Thanks for the update @JosepSampe . Those changes sound great.

Sergii is now taking over a lot of the Lithops-related work on our side. @sergii-mamedov FYI please see Josep's comment above.

**Comment by sergii-mamedov** [comment]
Fixed after updating to lithops 2.5.x.
